app/lib/frontend/handlers/package.dart

```diff
@@ -249,6 +249,19 @@ Future<shelf.Response> _handlePackagePage({
   if (redirectPackageUrls.containsKey(packageName)) {
     return redirectResponse(redirectPackageUrls[packageName]!);
   }
+  // May redirect to the lowercased package name, but only if [versionName] and
+  // [assetKind] is unspecified.
+  // NOTE: we have legitimate mixed-case package names that must not be redirected.
+  final lowerCasePackageName = packageName.toLowerCase();
+  if (versionName == null &&
+      packageName != lowerCasePackageName &&
+      assetKind == AssetKind.readme) {
+    if (!await packageBackend.isPackageVisible(packageName) &&
```

Comment:
Is this lookup cheap?

Reply:
Yes, it should be: a redis-cached boolean flag.

Reply:
But it will mean an extra request in the common path - right?

Maybe better to put the check where we otherwise would have returned 404, so the cost is not paid in the common case.

Reply:
Ah - sorry you are right, I had not understood the full logic.

LGTM